File: fabric/src/main/java/dev/kineticcat/hexportation/fabric/transfer/ItemVariantImpl.java
package dev.kineticcat.hexportation.fabric.transfer;

import dev.kineticcat.hexportation.api.transfer.ItemVariant;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;

/**
 * Fabric implementation using singleton injection pattern.
 */
public class ItemVariantImpl extends ItemVariant.Implementation {
    
    public static final ItemVariantImpl INSTANCE = new ItemVariantImpl();
    
    @Override
    public ItemVariant of(Item item) {
        net.fabricmc.fabric.api.transfer.v1.item.ItemVariant fabricVariant = 
            net.fabricmc.fabric.api.transfer.v1.item.ItemVariant.of(item);
        return new FabricItemVariantWrapper(fabricVariant);
    }
    
    @Override
    public ItemVariant blank() {
        net.fabricmc.fabric.api.transfer.v1.item.ItemVariant fabricVariant = 
            net.fabricmc.fabric.api.transfer.v1.item.ItemVariant.blank();
        return new FabricItemVariantWrapper(fabricVariant);
    }
    
    
    /**
     * Wrap an existing Fabric ItemVariant into our wrapper.
     * Used by StorageView implementations.
     */
    public static ItemVariant wrap(net.fabricmc.fabric.api.transfer.v1.item.ItemVariant fabricVariant) {
        return new FabricItemVariantWrapper(fabricVariant);
    }
    
    /**
     * Wrapper class that extends our ItemVariant abstract class using Fabric's ItemVariant.
     */
    public static class FabricItemVariantWrapper extends ItemVariant {
        private final net.fabricmc.fabric.api.transfer.v1.item.ItemVariant fabricVariant;
        
        public FabricItemVariantWrapper(net.fabricmc.fabric.api.transfer.v1.item.ItemVariant fabricVariant) {
            this.fabricVariant = fabricVariant;
        }
        
        @Override
        public Item getItem() {
            return fabricVariant.getItem();
        }
        
        @Override
        public ItemStack toStack() {
            return fabricVariant.toStack();
        }
        
        @Override
        public ItemStack toStack(int count) {
            return fabricVariant.toStack(count);
        }
        
        public net.fabricmc.fabric.api.transfer.v1.item.ItemVariant getFabricVariant() {
            return fabricVariant;
        }
    }
}

File: fabric/src/main/java/dev/kineticcat/hexportation/fabric/transfer/ItemVariantImpl.kt
package dev.kineticcat.hexportation.fabric.transfer

import dev.kineticcat.hexportation.api.transfer.ItemVariant
import net.minecraft.world.item.Item
import net.minecraft.world.item.ItemStack
import net.fabricmc.fabric.api.transfer.v1.item.ItemVariant as FabricItemVariant

/**
 * Fabric implementation using singleton injection pattern.
 */
class ItemVariantImpl : ItemVariant.Implementation() {
    
    companion object {
        @JvmField
        val INSTANCE = ItemVariantImpl()
        
        /**
         * Wrap an existing Fabric ItemVariant into our wrapper.
         * Used by StorageView implementations.
         */
        @JvmStatic
        fun wrap(fabricVariant: FabricItemVariant): ItemVariant = 
            FabricItemVariantWrapper(fabricVariant)
    }
    
    override fun of(item: Item): ItemVariant = 
        FabricItemVariantWrapper(FabricItemVariant.of(item))
    
    override fun blank(): ItemVariant = 
        FabricItemVariantWrapper(FabricItemVariant.blank())
    
    /**
     * Wrapper class that extends our ItemVariant abstract class using Fabric's ItemVariant.
     */
    class FabricItemVariantWrapper(
        private val fabricVariant: FabricItemVariant
    ) : ItemVariant() {
        
        override val item: Item get() = fabricVariant.item
        
        override fun toStack(): ItemStack = fabricVariant.toStack()
        
        override fun toStack(count: Int): ItemStack = fabricVariant.toStack(count)
        
        fun getFabricVariant() = fabricVariant
    }
}